File: RGBEngine/src/main/java/engine/graphics/opengl/BufferObject.java
package engine.graphics.opengl;

import static org.lwjgl.opengl.GL15.*;

public class BufferObject extends GLObject {

    final int type;

    public BufferObject(int type, float[] data) {
        this(type);
        putData(data);
    }

    public BufferObject(int type, int[] data) {
        this(type);
        putData(data);
    }

    public BufferObject(int type) {
        super(glGenBuffers());
        this.type = type;
    }

    @Override
    public void bind() {
        GLState.bindBuffer(this);
    }

    @Override
    public void destroy() {
        glDeleteBuffers(id);
    }

    public final void putData(float[] data) {
        putData(data, GL_STATIC_DRAW);
    }

    public final void putData(float[] data, int usage) {
        bind();
        glBufferData(type, data, usage);
    }

    public final void putData(int[] data) {
        putData(data, GL_STATIC_DRAW);
    }

    public final void putData(int[] data, int usage) {
        bind();
        glBufferData(type, data, usage);
    }
}

File: Cookie/src/main/kotlin/engine/graphics/opengl/BufferObject.kt
package engine.graphics.opengl

import org.lwjgl.opengl.GL15.*

class BufferObject(internal val type: Int) : GLObject(glGenBuffers()) {

    constructor(type: Int, data: FloatArray) : this(type) {
        putData(data)
    }

    constructor(type: Int, data: IntArray) : this(type) {
        putData(data)
    }

    override fun bind() {
        GLState.bindBuffer(this)
    }

    override fun destroy() {
        glDeleteBuffers(id)
    }

    @JvmOverloads
    fun putData(data: FloatArray, usage: Int = GL_STATIC_DRAW) {
        bind()
        glBufferData(type, data, usage)
    }

    @JvmOverloads
    fun putData(data: IntArray, usage: Int = GL_STATIC_DRAW) {
        bind()
        glBufferData(type, data, usage)
    }
}
